Обратная сторона реальности

S-E-Pashin opened the pull request:
---
:mortar_board: [Обратная сторона реальности](https://up.htmlacademy.ru/react-individual/2/user/1072207/tasks/5)

S-E-Pashin committed:
Переустанавливает зависимость для vite. Создает моковые данные и типы для предложений и комментариев итерация 1. Передает данные в компонент/ы карточек предложений итерация 1.

S-E-Pashin committed:
Передает данные в компонент/ы карточек предложений итерация 2.
Создает компонент Перечня карточек предложений, добавляет функционал сохранения в свойство данного компонента id карточки на которую было произведено наведение.

Finnko commented on src/components/app/app.tsx:
```diff
@@ -8,16 +8,26 @@ import NotFound from 'pages/not-found/not-found.tsx';
-          element={<Main totalOffers={DataCities.TotalOffers}/>}
+          element={
+            <Main
+              // totalOffers={DataCities.TotalOffers}
```
Удаляем не надо закомментированный код хранить

Finnko commented on src/components/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,45 @@
+}
+
+function OffersList({offers}: OffersListProps): JSX.Element {
+  const [activeCard, saveIdActiveCard] = useState('Пусто');
```
Если начальное значение не определено то undefined или null 
```suggestion
  const [, saveIdActiveCard] = useState('Пусто');
```

Finnko replied:
Посмотри какой паттерн нейминга у useState надо строго соблюдать

Finnko commented on src/components/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,45 @@
+function OffersList({offers}: OffersListProps): JSX.Element {
+  const [activeCard, saveIdActiveCard] = useState('Пусто');
+
+  const cardArr: Card[] = [];
+  const numberCards: number = offers.length;
+  function getCardsArr(number: number) {
+    for (let i = 0; i < number; i++) {
+      cardArr.push(
+        {
+          id: offers[i].id,
+          content:
+            <PlaceCard offer = {offers[i]}
+              giveActiveCard={() => saveIdActiveCard(offers[i].id)}
+            />,
+        }
+      );
+    }
+  }
```
Это больше не надо так как нам сразу массив приходит https://react.dev/learn/rendering-lists

Finnko commented on src/components/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,45 @@
+  return (
+    <div className="cities__places-list places__list tabs__content">
+      {
+        cardArr.map((item) => <React.Fragment key={item.id}>{item.content}</React.Fragment>)
```
Фрагмент надо удалить

Finnko commented on src/components/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,45 @@
+          id: offers[i].id,
+          content:
+            <PlaceCard offer = {offers[i]}
+              giveActiveCard={() => saveIdActiveCard(offers[i].id)}
```
По критерию обработчики прокидываемые в компоненты именуем on + что то + действие onButtonClick

Finnko commented on src/components/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,45 @@
+  getCardsArr(numberCards);
+
+  return (
+    <div className="cities__places-list places__list tabs__content">
```
tabs__content можно удалить

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,31 @@
+}
+
+function PlaceCard({offer, giveActiveCard}: PlaceCardProps): JSX.Element {
+  const rating: number = offer.rating * 2;
```
Здесь надо считать пропорцию от общего количества звезд, на текущий момент очень неявное преобразование. Плюс по тз надо округлять

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,31 @@
-        <span>Premium</span>
-      </div>
+    <article className="cities__card place-card" onMouseOver={giveActiveCard}>
+      {offer.isPremium ? <div className="place-card__mark"><span>Premium</span></div> : null}
```
Используем conditional rendering, этот див в более больших проектах был бы компонентом беджа, следует вынести отдельно

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,31 @@
+    <article className="cities__card place-card" onMouseOver={giveActiveCard}>
+      {offer.isPremium ? <div className="place-card__mark"><span>Premium</span></div> : null}
       <div className="cities__image-wrapper place-card__image-wrapper">
         <a href="#">
```
Здесь должна быть ссылка роутера

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,31 @@
           <img
             className="place-card__image"
-            src="img/apartment-01.jpg"
+            src={offer.images[0]}
```
previewImage

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,31 @@
+            <b className="place-card__price-value">€{offer.price}</b>
             <span className="place-card__price-text">/&nbsp;night</span>
           </div>
           <button
```
На этой кнопке будет логика надо вынести отдельно

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -37,13 +44,13 @@ function PlaceCard(): JSX.Element {
         </div>
```
В кнопке текст To bookmarks меняется в зависимости от состояния

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -37,13 +44,13 @@ function PlaceCard(): JSX.Element {
         <div className="place-card__rating rating">
           <div className="place-card__stars rating__stars">
-            <span style={{width: '80%'}}/>
+            <span style={{width: `${rating}0%`}}/>
```
Очень неявно, не надо так делать

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -37,13 +44,13 @@ function PlaceCard(): JSX.Element {
           </div>
         </div>
         <h2 className="place-card__name">
           <a href="#">
```
Линк роутера, маршрут надо сформировать сразу актуальный. Смотри утилиту generatePath роутера

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -37,13 +44,13 @@ function PlaceCard(): JSX.Element {
+            {offer.title}
           </a>
         </h2>
         <p className="place-card__type">Apartment</p>
```
Тип берем из предложения

Finnko commented on src/index.tsx:
```diff
@@ -2,12 +2,18 @@ import React from 'react';
 root.render(
   <React.StrictMode>
-    <App />
+    <App
+      offers = {offers}
+      reviews = {reviews}
+    />
```
форматируем только так
```suggestion
    <App
      offers={offers}
      reviews={reviews}
    />
```

Finnko commented on src/mocks/offers.ts:
```diff
@@ -0,0 +1,186 @@
+  isPro: boolean;
+}
+
+export type OfferT = {
```
или просто Offer или TOffer

Finnko replied:
Для карточки в списке сервер отдает необогащенные данные надо завести под это тип

Finnko commented on src/mocks/offers.ts:
```diff
@@ -0,0 +1,186 @@
+}
+
+
+export const offers: OfferT[] = [
```
Стабы лучше перенести в папку mocks, скопировать ответ сервера

Finnko commented on src/pages/main/main.tsx:
```diff
@@ -1,30 +1,18 @@
-function Main({totalOffers}: MainProps): JSX.Element {
-  const cardArr: Card[] = [];
+function Main({offers}: MainProps): JSX.Element {
+  // todo возможно стоит передавать из листа предложений
+  // Если выполнять сортировку там то и количество предложений
+  // должно идти после сортировки
+  const numberCards: number = offers.length;
```
```suggestion
```

S-E-Pashin committed:
Создает отдельные моки для перечня и для карточки предложения. Передает данные в компонент/ы карточек предложений.
Создает компонент Перечня карточек предложений, добавляет функционал сохранения в свойство данного компонента id карточки на которую было произведено наведение. Сегментирует header и logo. Лого в зависимости от установленного модификатора из макета. Вносит обусловленные ранее корректировки.

Finnko commented on src/components/header/header.tsx:
```diff
@@ -0,0 +1,36 @@
+      <div className="container">
+        <div className="header__wrapper">
+          <div className="header__left">
+            <Logo blockName={'header'} active/>
```
Для строк не нужна интерполяция
```suggestion
            <Logo blockName="header" active/>
```

Finnko commented on src/components/logo/logo.tsx:
```diff
@@ -0,0 +1,36 @@
+
+function Logo({blockName, active}: LogoProps): JSX.Element {
+  const size = sizes[blockName];
+  const additionalClass = active ? (`${blockName}__logo-link ${blockName}__logo-link--active`) : `${blockName}__logo-link`;
```
Это надо сделать через clsх/classnames как в компоненте карты

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,24 @@
-  giveActiveCard: () => void;
+type PlaceCardProps={
+  offer: OfferFromList;
+  setCard: () => void;
```
коллбек передаваемый пропс on + что то

Finnko commented on src/components/place-card/place-card.tsx:
```diff
@@ -1,24 +1,24 @@
 
   return (
-    <article className="cities__card place-card" onMouseOver={giveActiveCard}>
+    <article className="cities__card place-card" onMouseOver={setCard}>
```
Когда я мышку увожу с карточки значение в стейте становится снов null

S-E-Pashin committed:
Вносит использование cn.
Переносит типы моков в отдельный раздел.
Убирает интерполяцию в строках.
Изменяет наименование сеттера в карточке объявления для главной страницы.
Добавляет обнуление стейта наведения на карточку для перечня на главной странице.
Избавляется от магических чисел в компоненте рейтинга. Добавляет возможность использования различных систем исчисления, как 5 так и 10и бальной системы.
Выносит Бэйдж премиум в отдельный компонент.
Заменяет в карточке офера главной страницы применение тернарного оператора на conditional rendering.
Заменяет в карточке ссылку на link для переадресации на страницу офера для данной карточки.
Выносит кнопку закладки картоки офера в отдельный компонент.
Выносит карту отдельным компонентом.
Применяет лайаут более сегментированного использования компонентов.
Вводит применения generatePath для путей карточек оферов.
Производит дополнительное сегментирование табов, сортировки, формы и избранного.

S-E-Pashin committed:
Поднимает раздел кнопки.
Деструктурирует пропс.
Стейт в форме из 2х сводит в один объект.

S-E-Pashin committed:
Резервное копирование!

S-E-Pashin committed:
Добавляет интерактивность на страницу избранного, группирует и отображает оферы по городам. Использует один компонент карточки объявления для отрисовки на главной и на странице избранного.

S-E-Pashin committed:
Итерация 11.04

S-E-Pashin committed:
Изменяет порядок обработки сортировки по городу для оферов в избранном.
